### PR TITLE
implement FormData.get and FormData.getAll

### PR DESCRIPTION
**Pull request by jdeniau**
The implementation is maybe very simplistic but it may do the job

See:

* [get](https://xhr.spec.whatwg.org/#dom-formdata-get)
* [getAll](https://xhr.spec.whatwg.org/#dom-formdata-getall)

### CODE REVIEW

**Comment by jdeniau** [comment]
Th appveyor test is failing randomly I don't know why (I tried to amend and push another time, and it failed on another test case)

**Comment by yordis** [comment]
@alexindigo could you take a look to this mate. I need this feature/

File: lib/form_data.js [hunk 1]
```diff
@@ -31,6 +31,7 @@ function FormData(options) {
   this._overheadLength = 0;
   this._valueLength = 0;
   this._valuesToMeasure = [];
+  this._entryList = [];
```
Comment (yordis) [inline lib/form_data.js]: I am confused with this.

You use an array as the data structure but then you use it as an object

Comment (jdeniau) [reply lib/form_data.js]: My bad ! 
It works because Javascript Array are objects, and thus it become attribute of an object, I will change this to a map ;)

Comment (yordis) [reply lib/form_data.js]: @jdeniau actually I would use a list for this and create a list of objects.

I know that from data structure makes sense but following the spec I believe that `append` do not remove the previous one.

From documentation


> Each FormData object has an associated entry list (a list of entries). It is initially the empty list.
> An entry consists of a name and a value.


So maybe is better to have a  list of objects

```
[
 { name: 'hello', value: 'world'}
]
```

I am trying to finish the spec and probably maybe would be better if you change the current implementation

So far,

```
function createEntry(name, value, filename) {
  // TODO: handle Blob and File
  // Related to https://xhr.spec.whatwg.org/#concept-formdata-entry

  // Case 1:
  // If value is a Blob object and not a File object, then set value to a new
  // File object, representing the same bytes, whose name attribute value is
  // "blob".

  // Case 2:
  // If value is (now) a File object and filename is given, then set value to a
  // new File object, representing the same bytes, whose name attribute value
  // is filename.

  var entry = {
    name: name,
    value: value,
  };

  return entry;
}
```

I created this functions because it seems that we will need to do some complex stuff with the entries  https://xhr.spec.whatwg.org/#concept-formdata-entry

So basically for now you could do `this._entries.push( createEntry(name, value, filename  );`

File: lib/form_data.js [hunk 2]
```diff
@@ -31,6 +31,7 @@ function FormData(options) {
   this._overheadLength = 0;
   this._valueLength = 0;
   this._valuesToMeasure = [];
+  this._entryList = {};
```
Comment (yordis) [inline lib/form_data.js]: We need to use a list rather than an array. The current spec allow you to append to the list multiple values for the same `key`

File: lib/form_data.js [hunk 3]
```diff
@@ -76,6 +77,19 @@ FormData.prototype.append = function(field, value, options) {
 
   // pass along options.knownLength
   this._trackLength(header, value, options);
+
+  if (!this._entryList[field]) {
+    this._entryList[field] = [];
+  }
+  this._entryList[field].push(value);
```
Comment (yordis) [inline lib/form_data.js]: Append an object with the following shape

```
{ name: '', value: '' }
```

`append` do not remove previous entry, just do an `append` on the list.

File: lib/form_data.js [hunk 4]
```diff
@@ -76,6 +77,19 @@ FormData.prototype.append = function(field, value, options) {
...
+};
+
+FormData.prototype.get = function(field) {
+  return this._entryList[field] ? this._entryList[field][0] : null;
```
Comment (yordis) [inline lib/form_data.js]: From the spec

> The get(name) method, when invoked, must return the value of the first entry whose name is `name` from the context object’s entry list, and null otherwise.


Do an `Array.find`

File: lib/form_data.js [hunk 5]
```diff
@@ -76,6 +77,19 @@ FormData.prototype.append = function(field, value, options) {
...
+};
+
+FormData.prototype.getAll = function(field) {
+  return this._entryList[field] || [];
```
Comment (yordis) [inline lib/form_data.js]: From the spec
> The getAll(name) method, when invoked, must return the values of all entries whose name is name, in order, from the context object’s entry list, and the empty list otherwise.

Do a `Array.filter` in the list.

**Comment by yordis** [comment]
Hey @jdeniau, for me to be able to finish the other APIs I will need the changes that I suggest. Would be better for me to work on top what you did already rather than duplicating the work.

**Comment by jdeniau** [comment]
Looks good to me, I will work on this, but I did not quite well understand the [constructing the entry list(https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constructing-the-form-data-set) part :thinking:

**Comment by yordis** [comment]
@jdeniau well that one is for the web actually

`FormData` constructor expects an optional `form`.

If you pass a form you need to loop throw the list of inputs under that form and do an `append` of the entry where the entry name is the field name and the value is the field value based on some weird stuff.

But this code is only for NodeJS, unless you are trying to monkey patch that as well.

File: lib/form_data.js [hunk 6]
```diff
@@ -76,6 +104,31 @@ FormData.prototype.append = function(field, value, options) {
...
+
+  return foundItem.value;
+};
+
+FormData.prototype.getAll = function(field) {
+  return this._entryList
+    .filter(function(item) {
+      return item.name === field;
+    })
+    .map(function(item) {
```
Comment (yordis) [inline lib/form_data.js]: Using `reduce` you avoid the extra loop.

File: lib/form_data.js [hunk 7]
```diff
@@ -76,6 +104,31 @@ FormData.prototype.append = function(field, value, options) {
...
+
+  this._entryList.push(createEntry(field, value));
+};
+
+FormData.prototype.get = function(field) {
+  var foundItem = this._entryList.find(function(item) {
+    return item.name === field;
+  });
+
+  if (!foundItem) {
```
Comment (yordis) [inline lib/form_data.js]: Just do `return foundItem ? foundItem.value : null`
